2017 - Prelim Solutions

shah-smit opened the pull request.

shah-smit committed:
added 2014 prelim folder with empty readme

shah-smit committed:
Merge branch 'master' of https://github.com/shah-smit/CO1109-Answers

shah-smit committed:
Update readme.md

shah-smit committed:
Update readme.md

shah-smit committed:
added prelim folder

shah-smit committed:
Update readme.md

shah-smit committed:
Update readme.md

shah-smit committed:
Update readme.md

shrmnk commented on 2017-prelim/readme.md:
```diff
@@ -0,0 +1,309 @@
+
+* c) i) 8
+* c) ii) It will throw error, InputMismatchException
+* c) iii) Try{} Catch(Exception ex){}
```
Is the capitalised 't' in Try and 'c' in Catch on purpose?

shah-smit replied:
its a typo. Will change it.

shrmnk commented on 2017-prelim/readme.md:
````diff
@@ -0,0 +1,309 @@
+```java
+void printChars(int start, int end, String s){
+  int len = s.length()-1;
+  if(len > 0 && start >= len && end <= s)
````
Did this compile? `end` is type `int` while `s` is type `String`.

shrmnk replied:
Also, did the question specify whether a null-check was required for `s`?

shrmnk commented on 2017-prelim/readme.md:
```diff
@@ -0,0 +1,309 @@
+      double sum = a[i][j] + b[i][j];
+      if(sum != s[i][j])
+      {
+        result false;
```
Did this compile? This statement does not seem right.

shrmnk commented on 2017-prelim/readme.md:
````diff
@@ -0,0 +1,309 @@
+
+* d)
+```java
+public static void playLines(String filename)throws Exception
````
It is generally bad practice to just throw a broad `Exception`. Did the question specifically choose this? Otherwise, please throw the specific Exceptions (e.g. `FileNotFoundException`).

shah-smit replied:
The header of the method was provided to us! I agree with you. But i need to follow the question paper format!

shrmnk commented on 2017-prelim/readme.md:
````diff
@@ -0,0 +1,309 @@
+```java
+public static void playLines(String filename)throws Exception
+{
+  Scanner in = new Scanner(new FileReader(filename));
````
Is the FileReader handler closed? You may want to consider using a [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) since [FileReader](https://docs.oracle.com/javase/7/docs/api/java/io/FileReader.html) implements [AutoCloseable](https://docs.oracle.com/javase/7/docs/api/java/lang/AutoCloseable.html)

shah-smit replied:
The header of the method was provided to us! I agree with you. But i need to follow the question paper format!

shrmnk commented on 2017-prelim/readme.md:
````diff
@@ -0,0 +1,309 @@
+```java
+Fraction(int numer, int denom)
+{
+  this.numer = number;
````
Did this compile? `number` does not exist

shrmnk commented on 2017-prelim/readme.md:
````diff
@@ -0,0 +1,309 @@
+```java
+String toString()
+{
+  String output = numer +"/"+ demon;
````
While I understand you may be doing this under a time-tested scenario, but since this repo is meant to be a reference answer bank, please use relevant variables names (or is this a typographical error? `demon` sounds like an unlikely variable name)

shah-smit replied:
Sure, will change it.

shrmnk commented on 2017-prelim/readme.md:
````diff
@@ -0,0 +1,309 @@
+```java
+boolean equals(Fraction f)
+{
+  if(f.number == numer && f.demon == demon)
````
I am assuming `numer` and `demon` are static or instance variables declared in the class. However, please ensure that the variable names are relevant.

shah-smit replied:
They are instance variables. It was mention in the question. We were not expected to write whole code. just the snippet.

shrmnk commented on 2017-prelim/readme.md:
```diff
@@ -0,0 +1,309 @@
+{
+  if(f.number == numer && f.demon == demon)
+  {
+    return true;
```
It seems this entire method body can be compressed into 1 line.

i.e. `return f.number == numer && f.demon == demon;`

(Assuming your variable names are correct)

shah-smit replied:
@shrmnk I know we can implement that. But I am trying to write code that can be read by newbies.

shrmnk commented on 2017-prelim/readme.md:
````diff
@@ -0,0 +1,309 @@
+}
+```
+
+Note: All the above answers may have one or many solutions. And di confirm each by compiling it.
````
Minor typographical mistake here, please correct!

shrmnk reviewed (changes requested):
Please also be consistent in the formatting of your code, especially in the locations of your opening braces `{`. Some of your braces appear in-line while others appear on new lines. It would be preferable to keep consistent with the rest of the repository by keeping opening braces in-line. i.e.
`for(int i = 0; i < 3; i++) {`

shrmnk commented:
Generally, the solutions in the repo which can be tested by writing and compiling are done before the answers are uploaded. Do try and compile them when you have the time!

shah-smit commented:
Will check them out!

shah-smit committed:
Update readme.md

shah-smit committed:
Changed the format

Changed the format of the code such that it follows the syntax of block starting inline with method name
